sql/core/src/main/scala/org/apache/spark/sql/execution/ShuffledRowRDD.scala

```diff
@@ -177,19 +177,36 @@ class ShuffledRowRDD(
     val tracker = SparkEnv.get.mapOutputTracker.asInstanceOf[MapOutputTrackerMaster]
     partition.asInstanceOf[ShuffledRowRDDPartition].spec match {
       case CoalescedPartitionSpec(startReducerIndex, endReducerIndex, _) =>
-        // TODO order by partition size.
-        startReducerIndex.until(endReducerIndex).flatMap { reducerIndex =>
-          tracker.getPreferredLocationsForShuffle(dependency, reducerIndex)
-        }
+        tracker.getPreferredLocationsForShuffle(
+          dependency,
+          0,
+          dependency.rdd.getNumPartitions,
+          startReducerIndex,
+          endReducerIndex)
 
-      case PartialReducerPartitionSpec(_, startMapIndex, endMapIndex, _) =>
-        tracker.getMapLocation(dependency, startMapIndex, endMapIndex)
+      case PartialReducerPartitionSpec(reducerIndex, startMapIndex, endMapIndex, _) =>
+        tracker.getPreferredLocationsForShuffle(
```

Comment:
Wouldn't this change the behavior? Previously, `getMapLocation` returns the location anyway but`getPreferredLocationsForShuffle` seems to only return the location only when it has large shuffle data size.

Reply:
According to the original idea of preferred locations. It's more effective to schedule task to a bigger data size node. And same explanation with the comment of the threshold `REDUCER_PREF_LOCS_FRACTION`.
```scala
// Fraction of total map output that must be at a location for it to considered as a preferred
// location for a reduce task. Making this larger will focus on fewer locations where most data
// can be read locally, but may lead to more delay in scheduling if those locations are busy.
```

If return locations of all blocks, the preferred locations will be less meaning since the location can acutally contain less data and we can only reduce a little data size with network traffic. One more probleam, store all locations at driver side will cause OOM, and it get worser if we allocate more executors.

So this pr just correct the behavior of preferred locations. But if you think it's worth to return more locations, we can expose the `REDUCER_PREF_LOCS_FRACTION` as a config (I think it should be another thread ?).

Hope it makes sense for you.

Reply:
My concern is whether this change would affect AQE perf since the `getMapLocation` was for the local shuffler reader originally. @cloud-fan Could you take a look at this?

Reply:
I agree. Now we have two cases of local shuffle reader:

- `PartialMapperPartitionSpec`,  if the partition < 1000, it has no affect since it only has one map and always return a location.
- `CoalescedMapperPartitionSpec `, it is a trade-off between small partition and data locality that takes some maps into on reducer. So I think it should follow this behavior.

Given this, seems we can relax the condition for the special case:
```scala
if (startMapIndex + 1 = endMapIndex)
  return that map location.
else
...
``` 

what do you think ? @Ngone51

Reply:
I think it's better to let AQE follow the battle-tested normal shuffle fetching logic. We can run some AQE benchmarks to gain more confidence though.